src/Model/RecordLocale.php

```diff
@@ -367,16 +375,31 @@ public function getSourceLocale(): ?Locale
     {
         /** @var DataObject|FluentExtension $record */
         $record = $this->getOriginalRecord();
+        $config = $record->config();
+
+        $isFrontend = FluentState::singleton()->getIsFrontend();
+        $inheritanceMode = $isFrontend
+            ? $config->get('frontend_publish_required')
+            : $config->get('cms_localisation_required');
 
+        // This model has localised data in the current locale so the current locale is also the source locale
         if ($record->existsInLocale($this->getLocale())) {
             return $this->getLocaleObject();
         }
 
+        // This model requires localisation so fallback of any kind is not allowed
+        // hence the content can't come from another locale
+        // We don't have a source locale for such case
+        if ($inheritanceMode === FluentExtension::INHERITANCE_MODE_EXACT) {
```

Comment:
Looks good! I was trying to remind myself how inheritance worked, but this line seems to obviously fix the issue. We should never care about fallbacks if inheritance is exact mode only.